In [1]:
from google.colab import drive

In [2]:
drive.mount('/content/drive')

Mounted at /content/drive


#  **ETL for Machine Learning Model 1**

Now, with the datasets unnested and organized, we are going to do some merges and data trasnformations in order to have the information required in a CSV file to develop these functions and models:

6. **Game_Recommendation**(`item_id`: int) Function:

  By entering the `item_id`, we should receive a list with 5 recommended games similar to the one entered.

  Example of return: [{"Rec 1": "PAYDAY™ The Heist"}, {"Rec 2": "The Kings' Crusade"}, {"Rec 3": "Red Faction®: Armageddon™"}, {"Rec 4": "No Man's Sky"}, {"Rec 5": "Robocraft"} ]


### 1. Data Load:

In [3]:
import pandas as pd
import numpy as np

In [4]:
df_steam_games = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/steam_games_final.csv')
df_steam_games

publisher                                             genres  \
0             Kotoshiro  ['Action', 'Casual', 'Indie', 'Simulation', 'S...   
1      Making Fun, Inc.       ['Free to Play', 'Indie', 'RPG', 'Strategy']   
2          Poolians.com  ['Casual', 'Free to Play', 'Indie', 'Simulatio...   
3                  彼岸领域                  ['Action', 'Adventure', 'Casual']   
4                   NaN                                                NaN   
...                 ...                                                ...   
32130   Ghost_RUS Games      ['Casual', 'Indie', 'Simulation', 'Strategy']   
32131            Sacada                    ['Casual', 'Indie', 'Strategy']   
32132      Laush Studio                  ['Indie', 'Racing', 'Simulation']   
32133          SIXNAILS                                ['Casual', 'Indie']   
32134               NaN                                                NaN   

                       app_name                     title  \
0           Lost Summoner Kitty       Lost Summoner Kitty   
1                     Ironbound                 Ironbound   
2       Real Pool 3D - Poolians   Real Pool 3D - Poolians   
3                       弹炸人2222                   弹炸人2222   
4                 Log Challenge                       NaN   
...                         ...                       ...   
32130            Colony On Mars            Colony On Mars   
32131  LOGistICAL: South Africa  LOGistICAL: South Africa   
32132             Russian Roads             Russian Roads   
32133       EXIT 2 - Directions       EXIT 2 - Directions   
32134               Maze Run VR                       NaN   

                                                     url release_date  \
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
1      http://store.steampowered.com/app/643980/Ironb...   2018-01-04   
2      http://store.steampowered.com/app/670290/Real_...   2017-07-24   
3         http://store.steampowered.com/app/767400/2222/   2017-12-07   
4      http://store.steampowered.com/app/773570/Log_C...          NaN   
...                                                  ...          ...   
32130  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
32131  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
32132  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
32133  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
32134  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                    tags  \
0      ['Strategy', 'Action', 'Indie', 'Casual', 'Sim...   
1      ['Free to Play', 'Strategy', 'Indie', 'RPG', '...   
2      ['Free to Play', 'Simulation', 'Sports', 'Casu...   
3                      ['Action', 'Adventure', 'Casual']   
4                ['Action', 'Indie', 'Casual', 'Sports']   
...                                                  ...   
32130      ['Strategy', 'Indie', 'Casual', 'Simulation']   
32131                    ['Strategy', 'Indie', 'Casual']   
32132                  ['Indie', 'Simulation', 'Racing']   
32133  ['Indie', 'Casual', 'Puzzle', 'Singleplayer', ...   
32134  ['Early Access', 'Adventure', 'Indie', 'Action...   

                                             reviews_url  \
0      http://steamcommunity.com/app/761140/reviews/?...   
1      http://steamcommunity.com/app/643980/reviews/?...   
2      http://steamcommunity.com/app/670290/reviews/?...   
3      http://steamcommunity.com/app/767400/reviews/?...   
4      http://steamcommunity.com/app/773570/reviews/?...   
...                                                  ...   
32130  http://steamcommunity.com/app/773640/reviews/?...   
32131  http://steamcommunity.com/app/733530/reviews/?...   
32132  http://steamcommunity.com/app/610660/reviews/?...   
32133  http://steamcommunity.com/app/658870/reviews/?...   
32134  http://steamcommunity.com/app/681550/reviews/?...   

                         

In [5]:
df_steam_games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32135 entries, 0 to 32134
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   publisher     24073 non-null  object 
 1   genres        28852 non-null  object 
 2   app_name      32133 non-null  object 
 3   title         30085 non-null  object 
 4   url           32135 non-null  object 
 5   release_date  30068 non-null  object 
 6   tags          31972 non-null  object 
 7   reviews_url   32133 non-null  object 
 8   specs         31465 non-null  object 
 9   price         30758 non-null  object 
 10  early_access  32135 non-null  bool   
 11  id            32133 non-null  float64
 12  developer     28836 non-null  object 
dtypes: bool(1), float64(1), object(11)
memory usage: 3.0+ MB


**Step 1:**

df_steam_games has a column called `release_date`. We need to extract the year of this release date and place it in a new column called `release_year`. We ar going to do something similar as in the step 1.

In [6]:
df_steam_games

publisher                                             genres  \
0             Kotoshiro  ['Action', 'Casual', 'Indie', 'Simulation', 'S...   
1      Making Fun, Inc.       ['Free to Play', 'Indie', 'RPG', 'Strategy']   
2          Poolians.com  ['Casual', 'Free to Play', 'Indie', 'Simulatio...   
3                  彼岸领域                  ['Action', 'Adventure', 'Casual']   
4                   NaN                                                NaN   
...                 ...                                                ...   
32130   Ghost_RUS Games      ['Casual', 'Indie', 'Simulation', 'Strategy']   
32131            Sacada                    ['Casual', 'Indie', 'Strategy']   
32132      Laush Studio                  ['Indie', 'Racing', 'Simulation']   
32133          SIXNAILS                                ['Casual', 'Indie']   
32134               NaN                                                NaN   

                       app_name                     title  \
0           Lost Summoner Kitty       Lost Summoner Kitty   
1                     Ironbound                 Ironbound   
2       Real Pool 3D - Poolians   Real Pool 3D - Poolians   
3                       弹炸人2222                   弹炸人2222   
4                 Log Challenge                       NaN   
...                         ...                       ...   
32130            Colony On Mars            Colony On Mars   
32131  LOGistICAL: South Africa  LOGistICAL: South Africa   
32132             Russian Roads             Russian Roads   
32133       EXIT 2 - Directions       EXIT 2 - Directions   
32134               Maze Run VR                       NaN   

                                                     url release_date  \
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
1      http://store.steampowered.com/app/643980/Ironb...   2018-01-04   
2      http://store.steampowered.com/app/670290/Real_...   2017-07-24   
3         http://store.steampowered.com/app/767400/2222/   2017-12-07   
4      http://store.steampowered.com/app/773570/Log_C...          NaN   
...                                                  ...          ...   
32130  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
32131  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
32132  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
32133  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
32134  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                    tags  \
0      ['Strategy', 'Action', 'Indie', 'Casual', 'Sim...   
1      ['Free to Play', 'Strategy', 'Indie', 'RPG', '...   
2      ['Free to Play', 'Simulation', 'Sports', 'Casu...   
3                      ['Action', 'Adventure', 'Casual']   
4                ['Action', 'Indie', 'Casual', 'Sports']   
...                                                  ...   
32130      ['Strategy', 'Indie', 'Casual', 'Simulation']   
32131                    ['Strategy', 'Indie', 'Casual']   
32132                  ['Indie', 'Simulation', 'Racing']   
32133  ['Indie', 'Casual', 'Puzzle', 'Singleplayer', ...   
32134  ['Early Access', 'Adventure', 'Indie', 'Action...   

                                             reviews_url  \
0      http://steamcommunity.com/app/761140/reviews/?...   
1      http://steamcommunity.com/app/643980/reviews/?...   
2      http://steamcommunity.com/app/670290/reviews/?...   
3      http://steamcommunity.com/app/767400/reviews/?...   
4      http://steamcommunity.com/app/773570/reviews/?...   
...                                                  ...   
32130  http://steamcommunity.com/app/773640/reviews/?...   
32131  http://steamcommunity.com/app/733530/reviews/?...   
32132  http://steamcommunity.com/app/610660/reviews/?...   
32133  http://steamcommunity.com/app/658870/reviews/?...   
32134  http://steamcommunity.com/app/681550/reviews/?...   

                         

As in the step 1, it is a good practice to check the NaN in the column 'release_date'.

In [7]:
mask = df_steam_games['release_date'].isna()
df_steam_games[mask]

publisher                                            genres  \
4                   NaN                                               NaN   
11                  NaN                                               NaN   
19                  NaN                                               NaN   
20                  NaN                                               NaN   
22                  NaN                                               NaN   
...                 ...                                               ...   
32077               NaN                                               NaN   
32084               NaN                     ['Action', 'Casual', 'Indie']   
32085               NaN                                               NaN   
32088  rechargecomplete  ['Action', 'Adventure', 'Indie', 'Early Access']   
32134               NaN                                               NaN   

                   app_name              title  \
4             Log Challenge                NaN   
11     Icarus Six Sixty Six                NaN   
19            After Life VR                NaN   
20               Kitty Hawk                NaN   
22               Mortars VR                NaN   
...                     ...                ...   
32077           SpaceWalker                NaN   
32084              BOMBFEST           BOMBFEST   
32085            LIV Client                NaN   
32088     RECHARGE COMPLETE  RECHARGE COMPLETE   
32134           Maze Run VR                NaN   

                                                     url release_date  \
4      http://store.steampowered.com/app/773570/Log_C...          NaN   
11     http://store.steampowered.com/app/724910/Icaru...          NaN   
19     http://store.steampowered.com/app/772590/After...          NaN   
20     http://store.steampowered.com/app/640250/Kitty...          NaN   
22     http://store.steampowered.com/app/711440/Morta...          NaN   
...                                                  ...          ...   
32077  http://store.steampowered.com/app/705860/Space...          NaN   
32084  http://store.steampowered.com/app/729640/BOMBF...          NaN   
32085  http://store.steampowered.com/app/755540/LIV_C...          NaN   
32088  http://store.steampowered.com/app/708070/RECHA...          NaN   
32134  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                    tags  \
4                ['Action', 'Indie', 'Casual', 'Sports']   
11                                            ['Casual']   
19                       ['Early Access', 'Indie', 'VR']   
20     ['Early Access', 'Action', 'Adventure', 'Indie...   
22     ['Early Access', 'Strategy', 'Action', 'Indie'...   
...                                                  ...   
32077                         ['Early Access', 'Casual']   
32084                      ['Casual', 'Indie', 'Action']   
32085  ['Video Production', 'Utilities', 'Web Publish...   
32088   ['Early Access', 'Action', 'Adventure', 'Indie']   
32134  ['Early Access', 'Adventure', 'Indie', 'Action...   

                                             reviews_url  \
4      http://steamcommunity.com/app/773570/reviews/?...   
11     http://steamcommunity.com/app/724910/reviews/?...   
19     http://steamcommunity.com/app/772590/reviews/?...   
20     http://steamcommunity.com/app/640250/reviews/?...   
22     http://steamcommunity.com/app/711440/reviews/?...   
...                                                  ...   
32077  http://steamcommunity.com/app/705860/reviews/?...   
32084  http://steamcommunity.com/app/729640/reviews/?...   
32085  http://steamcommunity.com/app/755540/reviews/?...   
32088  http://steamcommunity.com/app/708070/reviews/?...   
32134  http://steamcommunity.com/app/681550/reviews/?...   

                                                   specs price  early_access  \
4      ['Single-player', 'Full controller support', '...  2.99         False   
11     ['

In this case, we can see that though the 'release_date' is in NaN, the entries have other information that can be useful.

Now, let's review the format of the dates in the column 'release_date', to check how can we extract the year.

In [8]:
df_steam_games['release_date'][:60]

0     2018-01-04
1     2018-01-04
2     2017-07-24
3     2017-12-07
4            NaN
5     2018-01-04
6     2018-01-04
7     2018-01-04
8     2018-01-04
9     2018-01-04
10        Soon..
11           NaN
12    2018-01-04
13    2018-01-04
14    2018-01-03
15    2018-01-03
16    2018-01-03
17    2018-01-03
18    2018-01-03
19           NaN
20           NaN
21    2018-01-03
22           NaN
23    2017-12-22
24    2017-12-23
25    2018-01-03
26    2018-01-03
27    1997-06-30
28    1998-11-08
29    2016-11-25
30           NaN
31    2018-01-01
32    2018-01-01
33    2018-01-01
34    2018-01-01
35    2017-12-30
36           NaN
37    2018-01-01
38    2018-01-01
39    2006-07-06
40    2006-07-06
41    2006-07-11
42           NaN
43          2017
44     Beta测试已开启
45           NaN
46    2017-12-29
47    2018-03-30
48    2017-12-29
49    2017-12-29
50    2017-12-29
51           NaN
52    2017-12-30
53           NaN
54    2017-12-29
55    2005-08-09
56    2006-09-29
57    2006-11-20
58    2006-11-

In the sample of the first 60 rows we can see that the normal format is YYYY-MM-DD. However, there are some NaN and strings that need to be considered. Also, there are entries with only the year. So we need to consider also this.

In [9]:
import re

# We create a function to extract the year considering the data that we have in the sample above.
def extract_year(date_str):
    if pd.notna(date_str):  #If the value is not NaN
        year_match = re.search(r'\b\d{4}\b', date_str) # We loo for the year following the format '\b\d{4}\b' using the function re.search()
        if year_match:
            return year_match.group()
    return np.NaN

# we apply the function to the column release_date
release_year = df_steam_games['release_date'].apply(extract_year)

# Mostrar el DataFrame resultante
release_year

0        2018
1        2018
2        2017
3        2017
4         NaN
         ... 
32130    2018
32131    2018
32132    2018
32133    2017
32134     NaN
Name: release_date, Length: 32135, dtype: object

Now we concatenate the column release_year with the dataframe steam_games:

In [10]:
df_steam_games_WY = pd.concat([df_steam_games, pd.DataFrame({'release_year': release_year})], axis=1)
df_steam_games_WY

publisher                                             genres  \
0             Kotoshiro  ['Action', 'Casual', 'Indie', 'Simulation', 'S...   
1      Making Fun, Inc.       ['Free to Play', 'Indie', 'RPG', 'Strategy']   
2          Poolians.com  ['Casual', 'Free to Play', 'Indie', 'Simulatio...   
3                  彼岸领域                  ['Action', 'Adventure', 'Casual']   
4                   NaN                                                NaN   
...                 ...                                                ...   
32130   Ghost_RUS Games      ['Casual', 'Indie', 'Simulation', 'Strategy']   
32131            Sacada                    ['Casual', 'Indie', 'Strategy']   
32132      Laush Studio                  ['Indie', 'Racing', 'Simulation']   
32133          SIXNAILS                                ['Casual', 'Indie']   
32134               NaN                                                NaN   

                       app_name                     title  \
0           Lost Summoner Kitty       Lost Summoner Kitty   
1                     Ironbound                 Ironbound   
2       Real Pool 3D - Poolians   Real Pool 3D - Poolians   
3                       弹炸人2222                   弹炸人2222   
4                 Log Challenge                       NaN   
...                         ...                       ...   
32130            Colony On Mars            Colony On Mars   
32131  LOGistICAL: South Africa  LOGistICAL: South Africa   
32132             Russian Roads             Russian Roads   
32133       EXIT 2 - Directions       EXIT 2 - Directions   
32134               Maze Run VR                       NaN   

                                                     url release_date  \
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
1      http://store.steampowered.com/app/643980/Ironb...   2018-01-04   
2      http://store.steampowered.com/app/670290/Real_...   2017-07-24   
3         http://store.steampowered.com/app/767400/2222/   2017-12-07   
4      http://store.steampowered.com/app/773570/Log_C...          NaN   
...                                                  ...          ...   
32130  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
32131  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
32132  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
32133  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
32134  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                    tags  \
0      ['Strategy', 'Action', 'Indie', 'Casual', 'Sim...   
1      ['Free to Play', 'Strategy', 'Indie', 'RPG', '...   
2      ['Free to Play', 'Simulation', 'Sports', 'Casu...   
3                      ['Action', 'Adventure', 'Casual']   
4                ['Action', 'Indie', 'Casual', 'Sports']   
...                                                  ...   
32130      ['Strategy', 'Indie', 'Casual', 'Simulation']   
32131                    ['Strategy', 'Indie', 'Casual']   
32132                  ['Indie', 'Simulation', 'Racing']   
32133  ['Indie', 'Casual', 'Puzzle', 'Singleplayer', ...   
32134  ['Early Access', 'Adventure', 'Indie', 'Action...   

                                             reviews_url  \
0      http://steamcommunity.com/app/761140/reviews/?...   
1      http://steamcommunity.com/app/643980/reviews/?...   
2      http://steamcommunity.com/app/670290/reviews/?...   
3      http://steamcommunity.com/app/767400/reviews/?...   
4      http://steamcommunity.com/app/773570/reviews/?...   
...                                                  ...   
32130  http://steamcommunity.com/app/773640/reviews/?...   
32131  http://steamcommunity.com/app/733530/reviews/?...   
32132  http://steamcommunity.com/app/610660/reviews/?...   
32133  http://steamcommunity.com/app/658870/reviews/?...   
32134  http://steamcommunity.com/app/681550/reviews/?...   

                         

Now we need to change the name of the column 'id' by 'item_id' to have a standard.

In [11]:
df_steam_games_WY.rename(columns={'id': 'item_id'}, inplace=True)
df_steam_games_WY

publisher                                             genres  \
0             Kotoshiro  ['Action', 'Casual', 'Indie', 'Simulation', 'S...   
1      Making Fun, Inc.       ['Free to Play', 'Indie', 'RPG', 'Strategy']   
2          Poolians.com  ['Casual', 'Free to Play', 'Indie', 'Simulatio...   
3                  彼岸领域                  ['Action', 'Adventure', 'Casual']   
4                   NaN                                                NaN   
...                 ...                                                ...   
32130   Ghost_RUS Games      ['Casual', 'Indie', 'Simulation', 'Strategy']   
32131            Sacada                    ['Casual', 'Indie', 'Strategy']   
32132      Laush Studio                  ['Indie', 'Racing', 'Simulation']   
32133          SIXNAILS                                ['Casual', 'Indie']   
32134               NaN                                                NaN   

                       app_name                     title  \
0           Lost Summoner Kitty       Lost Summoner Kitty   
1                     Ironbound                 Ironbound   
2       Real Pool 3D - Poolians   Real Pool 3D - Poolians   
3                       弹炸人2222                   弹炸人2222   
4                 Log Challenge                       NaN   
...                         ...                       ...   
32130            Colony On Mars            Colony On Mars   
32131  LOGistICAL: South Africa  LOGistICAL: South Africa   
32132             Russian Roads             Russian Roads   
32133       EXIT 2 - Directions       EXIT 2 - Directions   
32134               Maze Run VR                       NaN   

                                                     url release_date  \
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
1      http://store.steampowered.com/app/643980/Ironb...   2018-01-04   
2      http://store.steampowered.com/app/670290/Real_...   2017-07-24   
3         http://store.steampowered.com/app/767400/2222/   2017-12-07   
4      http://store.steampowered.com/app/773570/Log_C...          NaN   
...                                                  ...          ...   
32130  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
32131  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
32132  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
32133  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
32134  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                    tags  \
0      ['Strategy', 'Action', 'Indie', 'Casual', 'Sim...   
1      ['Free to Play', 'Strategy', 'Indie', 'RPG', '...   
2      ['Free to Play', 'Simulation', 'Sports', 'Casu...   
3                      ['Action', 'Adventure', 'Casual']   
4                ['Action', 'Indie', 'Casual', 'Sports']   
...                                                  ...   
32130      ['Strategy', 'Indie', 'Casual', 'Simulation']   
32131                    ['Strategy', 'Indie', 'Casual']   
32132                  ['Indie', 'Simulation', 'Racing']   
32133  ['Indie', 'Casual', 'Puzzle', 'Singleplayer', ...   
32134  ['Early Access', 'Adventure', 'Indie', 'Action...   

                                             reviews_url  \
0      http://steamcommunity.com/app/761140/reviews/?...   
1      http://steamcommunity.com/app/643980/reviews/?...   
2      http://steamcommunity.com/app/670290/reviews/?...   
3      http://steamcommunity.com/app/767400/reviews/?...   
4      http://steamcommunity.com/app/773570/reviews/?...   
...                                                  ...   
32130  http://steamcommunity.com/app/773640/reviews/?...   
32131  http://steamcommunity.com/app/733530/reviews/?...   
32132  http://steamcommunity.com/app/610660/reviews/?...   
32133  http://steamcommunity.com/app/658870/reviews/?...   
32134  http://steamcommunity.com/app/681550/reviews/?...   

                         

In [12]:
df_steam_games_WY.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32135 entries, 0 to 32134
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   publisher     24073 non-null  object 
 1   genres        28852 non-null  object 
 2   app_name      32133 non-null  object 
 3   title         30085 non-null  object 
 4   url           32135 non-null  object 
 5   release_date  30068 non-null  object 
 6   tags          31972 non-null  object 
 7   reviews_url   32133 non-null  object 
 8   specs         31465 non-null  object 
 9   price         30758 non-null  object 
 10  early_access  32135 non-null  bool   
 11  item_id       32133 non-null  float64
 12  developer     28836 non-null  object 
 13  release_year  29964 non-null  object 
dtypes: bool(1), float64(1), object(12)
memory usage: 3.2+ MB


Now we need to keep minimum columns that are requiered by the Machine Learning Model in order to save resources.

Analyzing the functions, the only columns that we are going to need are the following:

- item_id: It is important for the ML model where the user needs to enter an item_id in order to receive 5 games that are similar to the one selected.
- title: It is important for the ML model where the user needs to enter an item_id in order to receive 5 games that are similar to the one selected.
- release_year: This data can be useful because there are some gamers that prefer some vintage games better than the modern ones.
- genres: This can help us as a characterization of the video game.
- publisher: This can help us as a characterization of the video game.
- tags: This can help us as a characterization of the video game.
- specs: This can help us as a characterization of the video game.
- developer: This can help us as a characterization of the video game.

In [13]:
selected_columns = ['item_id','title','release_year','genres','publisher','tags','specs','developer']
df_FML1 = df_steam_games_WY[selected_columns]
df_FML1

item_id                     title release_year  \
0      761140.0       Lost Summoner Kitty         2018   
1      643980.0                 Ironbound         2018   
2      670290.0   Real Pool 3D - Poolians         2017   
3      767400.0                   弹炸人2222         2017   
4      773570.0                       NaN          NaN   
...         ...                       ...          ...   
32130  773640.0            Colony On Mars         2018   
32131  733530.0  LOGistICAL: South Africa         2018   
32132  610660.0             Russian Roads         2018   
32133  658870.0       EXIT 2 - Directions         2017   
32134  681550.0                       NaN          NaN   

                                                  genres         publisher  \
0      ['Action', 'Casual', 'Indie', 'Simulation', 'S...         Kotoshiro   
1           ['Free to Play', 'Indie', 'RPG', 'Strategy']  Making Fun, Inc.   
2      ['Casual', 'Free to Play', 'Indie', 'Simulatio...      Poolians.com   
3                      ['Action', 'Adventure', 'Casual']              彼岸领域   
4                                                    NaN               NaN   
...                                                  ...               ...   
32130      ['Casual', 'Indie', 'Simulation', 'Strategy']   Ghost_RUS Games   
32131                    ['Casual', 'Indie', 'Strategy']            Sacada   
32132                  ['Indie', 'Racing', 'Simulation']      Laush Studio   
32133                                ['Casual', 'Indie']          SIXNAILS   
32134                                                NaN               NaN   

                                                    tags  \
0      ['Strategy', 'Action', 'Indie', 'Casual', 'Sim...   
1      ['Free to Play', 'Strategy', 'Indie', 'RPG', '...   
2      ['Free to Play', 'Simulation', 'Sports', 'Casu...   
3                      ['Action', 'Adventure', 'Casual']   
4                ['Action', 'Indie', 'Casual', 'Sports']   
...                                                  ...   
32130      ['Strategy', 'Indie', 'Casual', 'Simulation']   
32131                    ['Strategy', 'Indie', 'Casual']   
32132                  ['Indie', 'Simulation', 'Racing']   
32133  ['Indie', 'Casual', 'Puzzle', 'Singleplayer', ...   
32134  ['Early Access', 'Adventure', 'Indie', 'Action...   

                                                   specs  \
0                                      ['Single-player']   
1      ['Single-player', 'Multi-player', 'Online Mult...   
2      ['Single-player', 'Multi-player', 'Online Mult...   
3                                      ['Single-player']   
4      ['Single-player', 'Full controller support', '...   
...                                                  ...   
32130            ['Single-player', 'Steam Achievements']   
32131  ['Single-player', 'Steam Achievements', 'Steam...   
32132  ['Single-player', 'Steam Achievements', 'Steam...   
32133  ['Single-player', 'Steam Achievements', 'Steam...   
32134  ['Single-player', 'Stats', 'Steam Leaderboards...   

                      developer  
0                     Kotoshiro  
1              Secret Level SRL  
2                  Poolians.com  
3                          彼岸领域  
4                           NaN  
...                         ...  
32130        Nikita "Ghost_RUS"  
32131                    Sacada  
32132  Laush Dmitriy Sergeevich  
32133             xropi,stev3ns  
32134                       NaN  

[32135 rows x 8 columns]

First, let's check if there is any item_id in NaN:

In [14]:
mask = df_FML1['item_id'].isna()
df_FML1[mask]

item_id                                           title release_year  \
74         NaN                                             NaN          NaN   
30961      NaN  Batman: Arkham City - Game of the Year Edition         2012   

                        genres  \
74                         NaN   
30961  ['Action', 'Adventure']   

                                               publisher  \
74                                                   NaN   
30961  Warner Bros. Interactive Entertainment, Feral ...   

                                                    tags  \
74                                                   NaN   
30961  ['Action', 'Open World', 'Batman', 'Adventure'...   

                                                   specs  \
74                                                   NaN   
30961  ['Single-player', 'Steam Achievements', 'Steam...   

                                        developer  
74                                            NaN  
30961  Rocksteady Studios,Feral Interactive (Mac)

Let's eliminate the item_id in NaN

In [15]:
df_FML1.dropna(subset=['item_id'], inplace=True)
df_FML1

<ipython-input-15-817a6f4b0782>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_FML1.dropna(subset=['item_id'], inplace=True)


item_id                     title release_year  \
0      761140.0       Lost Summoner Kitty         2018   
1      643980.0                 Ironbound         2018   
2      670290.0   Real Pool 3D - Poolians         2017   
3      767400.0                   弹炸人2222         2017   
4      773570.0                       NaN          NaN   
...         ...                       ...          ...   
32130  773640.0            Colony On Mars         2018   
32131  733530.0  LOGistICAL: South Africa         2018   
32132  610660.0             Russian Roads         2018   
32133  658870.0       EXIT 2 - Directions         2017   
32134  681550.0                       NaN          NaN   

                                                  genres         publisher  \
0      ['Action', 'Casual', 'Indie', 'Simulation', 'S...         Kotoshiro   
1           ['Free to Play', 'Indie', 'RPG', 'Strategy']  Making Fun, Inc.   
2      ['Casual', 'Free to Play', 'Indie', 'Simulatio...      Poolians.com   
3                      ['Action', 'Adventure', 'Casual']              彼岸领域   
4                                                    NaN               NaN   
...                                                  ...               ...   
32130      ['Casual', 'Indie', 'Simulation', 'Strategy']   Ghost_RUS Games   
32131                    ['Casual', 'Indie', 'Strategy']            Sacada   
32132                  ['Indie', 'Racing', 'Simulation']      Laush Studio   
32133                                ['Casual', 'Indie']          SIXNAILS   
32134                                                NaN               NaN   

                                                    tags  \
0      ['Strategy', 'Action', 'Indie', 'Casual', 'Sim...   
1      ['Free to Play', 'Strategy', 'Indie', 'RPG', '...   
2      ['Free to Play', 'Simulation', 'Sports', 'Casu...   
3                      ['Action', 'Adventure', 'Casual']   
4                ['Action', 'Indie', 'Casual', 'Sports']   
...                                                  ...   
32130      ['Strategy', 'Indie', 'Casual', 'Simulation']   
32131                    ['Strategy', 'Indie', 'Casual']   
32132                  ['Indie', 'Simulation', 'Racing']   
32133  ['Indie', 'Casual', 'Puzzle', 'Singleplayer', ...   
32134  ['Early Access', 'Adventure', 'Indie', 'Action...   

                                                   specs  \
0                                      ['Single-player']   
1      ['Single-player', 'Multi-player', 'Online Mult...   
2      ['Single-player', 'Multi-player', 'Online Mult...   
3                                      ['Single-player']   
4      ['Single-player', 'Full controller support', '...   
...                                                  ...   
32130            ['Single-player', 'Steam Achievements']   
32131  ['Single-player', 'Steam Achievements', 'Steam...   
32132  ['Single-player', 'Steam Achievements', 'Steam...   
32133  ['Single-player', 'Steam Achievements', 'Steam...   
32134  ['Single-player', 'Stats', 'Steam Leaderboards...   

                      developer  
0                     Kotoshiro  
1              Secret Level SRL  
2                  Poolians.com  
3                          彼岸领域  
4                           NaN  
...                         ...  
32130        Nikita "Ghost_RUS"  
32131                    Sacada  
32132  Laush Dmitriy Sergeevich  
32133             xropi,stev3ns  
32134                       NaN  

[32133 rows x 8 columns]

In [16]:
df_FML1.reset_index(drop=True, inplace=True)
df_FML1

item_id                     title release_year  \
0      761140.0       Lost Summoner Kitty         2018   
1      643980.0                 Ironbound         2018   
2      670290.0   Real Pool 3D - Poolians         2017   
3      767400.0                   弹炸人2222         2017   
4      773570.0                       NaN          NaN   
...         ...                       ...          ...   
32128  773640.0            Colony On Mars         2018   
32129  733530.0  LOGistICAL: South Africa         2018   
32130  610660.0             Russian Roads         2018   
32131  658870.0       EXIT 2 - Directions         2017   
32132  681550.0                       NaN          NaN   

                                                  genres         publisher  \
0      ['Action', 'Casual', 'Indie', 'Simulation', 'S...         Kotoshiro   
1           ['Free to Play', 'Indie', 'RPG', 'Strategy']  Making Fun, Inc.   
2      ['Casual', 'Free to Play', 'Indie', 'Simulatio...      Poolians.com   
3                      ['Action', 'Adventure', 'Casual']              彼岸领域   
4                                                    NaN               NaN   
...                                                  ...               ...   
32128      ['Casual', 'Indie', 'Simulation', 'Strategy']   Ghost_RUS Games   
32129                    ['Casual', 'Indie', 'Strategy']            Sacada   
32130                  ['Indie', 'Racing', 'Simulation']      Laush Studio   
32131                                ['Casual', 'Indie']          SIXNAILS   
32132                                                NaN               NaN   

                                                    tags  \
0      ['Strategy', 'Action', 'Indie', 'Casual', 'Sim...   
1      ['Free to Play', 'Strategy', 'Indie', 'RPG', '...   
2      ['Free to Play', 'Simulation', 'Sports', 'Casu...   
3                      ['Action', 'Adventure', 'Casual']   
4                ['Action', 'Indie', 'Casual', 'Sports']   
...                                                  ...   
32128      ['Strategy', 'Indie', 'Casual', 'Simulation']   
32129                    ['Strategy', 'Indie', 'Casual']   
32130                  ['Indie', 'Simulation', 'Racing']   
32131  ['Indie', 'Casual', 'Puzzle', 'Singleplayer', ...   
32132  ['Early Access', 'Adventure', 'Indie', 'Action...   

                                                   specs  \
0                                      ['Single-player']   
1      ['Single-player', 'Multi-player', 'Online Mult...   
2      ['Single-player', 'Multi-player', 'Online Mult...   
3                                      ['Single-player']   
4      ['Single-player', 'Full controller support', '...   
...                                                  ...   
32128            ['Single-player', 'Steam Achievements']   
32129  ['Single-player', 'Steam Achievements', 'Steam...   
32130  ['Single-player', 'Steam Achievements', 'Steam...   
32131  ['Single-player', 'Steam Achievements', 'Steam...   
32132  ['Single-player', 'Stats', 'Steam Leaderboards...   

                      developer  
0                     Kotoshiro  
1              Secret Level SRL  
2                  Poolians.com  
3                          彼岸领域  
4                           NaN  
...                         ...  
32128        Nikita "Ghost_RUS"  
32129                    Sacada  
32130  Laush Dmitriy Sergeevich  
32131             xropi,stev3ns  
32132                       NaN  

[32133 rows x 8 columns]

In the data provided we can check that there are some columns that can have other uses as a filtering for a better recommendation system, and not for training our model. Let's go deeper in this analysis:

1. **release_date:** We could potentially use the release_date column to filter recommendations based on the release date, so that we recommend newer or older games based on user preferences. For example, we might give more weight to games released in a similar time frame.

2. **publisher and developer:** We could potentially use the publisher and developer column to as filters in  order provide recommendations based on user preferences for specific publishers or developers. These columns tend to be very specific and it will not allow our model to generalize.

We need to remember that including additional columns in our recommendation system can make it more sophisticated and potentially improve the quality of recommendations. However, it also increases the complexity of our feature vectors and can make the vectorization process more challenging.

Now, we have these columns that are general properties that help us to describe better a videogame
1. gernes.
2. tags.
3. specs.

However, there are a lot of information inside of those columns, so the time and resources used for building the vectors, are going to be enormous. Taking this in consideration, we can check that the column 'tags' sometimes has information from the genres and the specs (Most characteristics to describe each item). So, to have a good model with the less of the resources consumption, we consider that using only the column 'tags' may be a good option.  

In [17]:
df_FML1_for_Model = df_FML1[['item_id','title','tags']]
df_FML1_for_Model

item_id                     title  \
0      761140.0       Lost Summoner Kitty   
1      643980.0                 Ironbound   
2      670290.0   Real Pool 3D - Poolians   
3      767400.0                   弹炸人2222   
4      773570.0                       NaN   
...         ...                       ...   
32128  773640.0            Colony On Mars   
32129  733530.0  LOGistICAL: South Africa   
32130  610660.0             Russian Roads   
32131  658870.0       EXIT 2 - Directions   
32132  681550.0                       NaN   

                                                    tags  
0      ['Strategy', 'Action', 'Indie', 'Casual', 'Sim...  
1      ['Free to Play', 'Strategy', 'Indie', 'RPG', '...  
2      ['Free to Play', 'Simulation', 'Sports', 'Casu...  
3                      ['Action', 'Adventure', 'Casual']  
4                ['Action', 'Indie', 'Casual', 'Sports']  
...                                                  ...  
32128      ['Strategy', 'Indie', 'Casual', 'Simulation']  
32129                    ['Strategy', 'Indie', 'Casual']  
32130                  ['Indie', 'Simulation', 'Racing']  
32131  ['Indie', 'Casual', 'Puzzle', 'Singleplayer', ...  
32132  ['Early Access', 'Adventure', 'Indie', 'Action...  

[32133 rows x 3 columns]

Now, we need to eliminate duplicates in order to have a better model:

In [18]:
df_FML1_for_Model.drop_duplicates(inplace=True)
df_FML1_for_Model

<ipython-input-18-8b41aaf2c067>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_FML1_for_Model.drop_duplicates(inplace=True)


item_id                     title  \
0      761140.0       Lost Summoner Kitty   
1      643980.0                 Ironbound   
2      670290.0   Real Pool 3D - Poolians   
3      767400.0                   弹炸人2222   
4      773570.0                       NaN   
...         ...                       ...   
32128  773640.0            Colony On Mars   
32129  733530.0  LOGistICAL: South Africa   
32130  610660.0             Russian Roads   
32131  658870.0       EXIT 2 - Directions   
32132  681550.0                       NaN   

                                                    tags  
0      ['Strategy', 'Action', 'Indie', 'Casual', 'Sim...  
1      ['Free to Play', 'Strategy', 'Indie', 'RPG', '...  
2      ['Free to Play', 'Simulation', 'Sports', 'Casu...  
3                      ['Action', 'Adventure', 'Casual']  
4                ['Action', 'Indie', 'Casual', 'Sports']  
...                                                  ...  
32128      ['Strategy', 'Indie', 'Casual', 'Simulation']  
32129                    ['Strategy', 'Indie', 'Casual']  
32130                  ['Indie', 'Simulation', 'Racing']  
32131  ['Indie', 'Casual', 'Puzzle', 'Singleplayer', ...  
32132  ['Early Access', 'Adventure', 'Indie', 'Action...  

[32132 rows x 3 columns]

We see that there's only one duplicate. Now, we are going to eliminate NaNs, considering that will not give value for our Recommendation System.

In [19]:
df_FML1_for_Model.dropna(inplace=True)
df_FML1_for_Model

<ipython-input-19-1488066a606c>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_FML1_for_Model.dropna(inplace=True)


item_id                     title  \
0      761140.0       Lost Summoner Kitty   
1      643980.0                 Ironbound   
2      670290.0   Real Pool 3D - Poolians   
3      767400.0                   弹炸人2222   
5      772540.0     Battle Royale Trainer   
...         ...                       ...   
32127  745400.0              Kebab it Up!   
32128  773640.0            Colony On Mars   
32129  733530.0  LOGistICAL: South Africa   
32130  610660.0             Russian Roads   
32131  658870.0       EXIT 2 - Directions   

                                                    tags  
0      ['Strategy', 'Action', 'Indie', 'Casual', 'Sim...  
1      ['Free to Play', 'Strategy', 'Indie', 'RPG', '...  
2      ['Free to Play', 'Simulation', 'Sports', 'Casu...  
3                      ['Action', 'Adventure', 'Casual']  
5      ['Action', 'Adventure', 'Simulation', 'FPS', '...  
...                                                  ...  
32127  ['Action', 'Indie', 'Casual', 'Violent', 'Adve...  
32128      ['Strategy', 'Indie', 'Casual', 'Simulation']  
32129                    ['Strategy', 'Indie', 'Casual']  
32130                  ['Indie', 'Simulation', 'Racing']  
32131  ['Indie', 'Casual', 'Puzzle', 'Singleplayer', ...  

[29922 rows x 3 columns]

Now, we have 29,922 entries without any NaN or duplicates. Finally, we need to check the type or our variables in the dataframe:

In [20]:
df_FML1_for_Model.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 29922 entries, 0 to 32131
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   item_id  29922 non-null  float64
 1   title    29922 non-null  object 
 2   tags     29922 non-null  object 
dtypes: float64(1), object(2)
memory usage: 935.1+ KB


We need to change the "title" to string:

In [21]:
df_FML1_for_Model['title'] = df_FML1_for_Model['title'].astype('string')
df_FML1_for_Model.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 29922 entries, 0 to 32131
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   item_id  29922 non-null  float64
 1   title    29922 non-null  string 
 2   tags     29922 non-null  object 
dtypes: float64(1), object(1), string(1)
memory usage: 935.1+ KB


<ipython-input-21-a82626cd68a0>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_FML1_for_Model['title'] = df_FML1_for_Model['title'].astype('string')


We are going to save the entire dataset in a CSV file to have a backup copy ready to be consumed in subsequent processes.

In [22]:
df_FML1_for_Model.to_csv('FML1.csv', index=False)

Let's save it also in a parquet file due to it saves resources.

In [23]:
df_FML1_for_Model.to_parquet('FML1.parquet', index=False)